File: clients/intellij/src/main/kotlin/dev/pubgrade/pub/PubDevApi.kt
package dev.pubgrade.pub

import com.google.gson.JsonParser
import java.net.URI
import java.net.URLEncoder
import java.net.http.HttpClient
import java.net.http.HttpRequest
import java.net.http.HttpResponse
import java.nio.charset.StandardCharsets
import java.time.Duration
import java.time.Instant
import java.util.concurrent.ConcurrentHashMap

/** Everything one pub.dev package document tells us. */
data class PubPackage(
    val latestVersion: String,
    /** Publish date per version, e.g. `1.2.3` -> Instant. */
    val publishedAt: Map<String, Instant>
)

private val REQUEST_TIMEOUT: Duration = Duration.ofSeconds(15)

/**
 * Reads pub.dev. One instance per project.
 *
 * Each package document is fetched at most once and then cached, because the
 * package list, the "outdated" check and the changelog dates all want the same
 * document. Call [clearCache] on refresh.
 */
class PubDevApi(
    private val http: HttpClient = HttpClient.newBuilder()
        .connectTimeout(REQUEST_TIMEOUT)
        .followRedirects(HttpClient.Redirect.NORMAL)
        .build()
) {
    // Values are wrapped because a ConcurrentHashMap cannot hold null, and
    // "we asked and pub.dev has no such package" is worth caching too.
    private val packages = ConcurrentHashMap<String, Box<PubPackage>>()
    private val changelogs = ConcurrentHashMap<String, Box<String>>()

    private class Box<T>(val value: T?)

    /** Null when the package is unknown or pub.dev is unreachable. */
    fun getPackage(name: String): PubPackage? =
        packages.computeIfAbsent(name) { Box(fetchPackage(it)) }.value

    /** Raw HTML of the changelog page. Null when it cannot be fetched. */
    fun getChangelogHtml(name: String): String? =
        changelogs.computeIfAbsent(name) {
            Box(get("https://pub.dev/packages/${encode(it)}/changelog"))
        }.value

    fun clearCache() {
        packages.clear()
        changelogs.clear()
    }

    private fun fetchPackage(name: String): PubPackage? {
        val body = get("https://pub.dev/api/packages/${encode(name)}") ?: return null

        val root = runCatching { JsonParser.parseString(body).asJsonObject }.getOrNull() ?: return null
        val latestVersion = root.getAsJsonObject("latest")
            ?.get("version")?.takeIf { it.isJsonPrimitive }?.asString
            ?: return null

        val publishedAt = LinkedHashMap<String, Instant>()
        root.getAsJsonArray("versions")?.forEach { element ->
            val entry = element.takeIf { it.isJsonObject }?.asJsonObject ?: return@forEach
            val version = entry.get("version")?.takeIf { it.isJsonPrimitive }?.asString ?: return@forEach
            val published = entry.get("published")?.takeIf { it.isJsonPrimitive }?.asString ?: return@forEach
            runCatching { Instant.parse(published) }.getOrNull()?.let { publishedAt[version] = it }
        }

        return PubPackage(latestVersion, publishedAt)
    }

    private fun get(url: String): String? = try {
        val request = HttpRequest.newBuilder(URI.create(url))
            .timeout(REQUEST_TIMEOUT)
            .header("User-Agent", "pubgrade-intellij")
            .GET()
            .build()

        val response = http.send(request, HttpResponse.BodyHandlers.ofString())
        if (response.statusCode() in 200..299) response.body() else null
    } catch (error: Exception) {
        null
    }

    private fun encode(value: String): String =
        URLEncoder.encode(value, StandardCharsets.UTF_8).replace("+", "%20")
}

File: clients/vscode/src/pub/pubDevApi.ts
import axios, { AxiosInstance } from 'axios';

/** Everything one pub.dev package document tells us. */
export interface PubPackage {
  latestVersion: string;
  /** Publish date per version, e.g. `1.2.3` -> Date. */
  publishedAt: Map<string, Date>;
}

const REQUEST_TIMEOUT_MS = 15_000;

/**
 * Reads pub.dev. One instance per extension activation.
 *
 * Each package document is fetched at most once and then cached, because the
 * package list, the "outdated" check and the changelog dates all want the same
 * document. Call `clearCache()` on refresh.
 */
export class PubDevApi {
  private readonly packages = new Map<string, Promise<PubPackage | null>>();
  private readonly changelogs = new Map<string, Promise<string | null>>();

  constructor(
    private readonly http: AxiosInstance = axios.create({
      timeout: REQUEST_TIMEOUT_MS,
      headers: { 'User-Agent': 'pubgrade-vscode' }
    })
  ) {}

  /** Null when the package is unknown or pub.dev is unreachable. */
  getPackage(name: string): Promise<PubPackage | null> {
    return this.remember(this.packages, name, () => this.fetchPackage(name));
  }

  /** Raw HTML of the changelog page. Null when it cannot be fetched. */
  getChangelogHtml(name: string): Promise<string | null> {
    return this.remember(this.changelogs, name, async () => {
      const url = `https://pub.dev/packages/${encodeURIComponent(name)}/changelog`;
      return this.get<string>(url);
    });
  }

  clearCache(): void {
    this.packages.clear();
    this.changelogs.clear();
  }

  private async fetchPackage(name: string): Promise<PubPackage | null> {
    const url = `https://pub.dev/api/packages/${encodeURIComponent(name)}`;
    const data = await this.get<any>(url);

    const latestVersion = data?.latest?.version;
    if (typeof latestVersion !== 'string') return null;

    const publishedAt = new Map<string, Date>();
    for (const entry of data.versions ?? []) {
      if (entry?.version && entry?.published) {
        publishedAt.set(entry.version, new Date(entry.published));
      }
    }

    return { latestVersion, publishedAt };
  }

  private async get<T>(url: string): Promise<T | null> {
    try {
      const response = await this.http.get<T>(url);
      return response.data;
    } catch (error) {
      console.error(`[Pubgrade] GET ${url} failed:`, error);
      return null;
    }
  }

  /** Caches the promise, not the value, so concurrent callers share one request. */
  private remember<T>(
    cache: Map<string, Promise<T>>,
    key: string,
    load: () => Promise<T>
  ): Promise<T> {
    const cached = cache.get(key);
    if (cached) return cached;

    const pending = load();
    cache.set(key, pending);
    return pending;
  }
}
